src/ostrowski/rationals/bounded.lean

```diff
@@ -0,0 +1,364 @@
+
+import analysis.special_functions.exp_log
+import analysis.special_functions.pow
+import data.real.cau_seq
+
+import ostrowski.rationals.basic
+import abvs_equiv
+
+open list is_absolute_value
+
+lemma prime_norm_lt_one_of_bounded_padic (abv : ℚ → ℝ) [habv : is_absolute_value abv]
+      (hnontriv: abv ≠ trivial_abs)
+      (all_nat_le_one: (∀ z : ℕ, abv z ≤ 1)):
+      ∃ p: ℕ, prime p ∧ abv p < 1 :=
+begin
+  by_contra,
+  push_neg at h,
+
+  suffices h': ∀ n: ℕ, n ≠ 0 → abv n = 1,
+  {
+    apply hnontriv,
+    set abv_hom := hom_of_abv abv with abv_hom_def,
+    set triv_hom := hom_of_abv trivial_abs with triv_hom_def,
+    have abv_hom_f: abv = abv_hom := by refl,
+    have triv_hom_f: trivial_abs = triv_hom := by refl,
+    rw [abv_hom_f, triv_hom_f],
+    apply mul_mor_eq_of_eq_on_pnat,
+    {
+      rw [← abv_hom_f, ← triv_hom_f],
+      rw [abv_neg abv, abv_neg trivial_abs],
+      rw [abv_one abv, abv_one trivial_abs],
+    },
+    {
+      intro n,
+      rw [← abv_hom_f, ← triv_hom_f],
+      unfold trivial_abs,
+      by_cases h'': n = 0,
+      {
+        rw h'',
+        simp,
+        rw abv_zero abv,
+      },
+      {
+        rw h' n h'',
+        simp [h''],
+      }
+    },
+  },
+  
+  apply @induction_on_primes (λ n, n ≠ 0 → abv n = 1),
+  { tauto },
+  { intro h, exact abv_one abv, },
+  {
+    intros p a p_prime ha hprod,
+    push_cast,
+    rw abv_mul abv,
+    have abvp_eq_one: abv p = 1,
+    { by linarith [h p p_prime, all_nat_le_one p], },
+    rw abvp_eq_one,
+    have a_ne_zero: a ≠ 0,
+    { by_contra h, push_neg at h, apply hprod, simp [h], },
+    rw ha a_ne_zero,
+    simp,
+  },
+end
+
+lemma padic_norm_q (p: ℕ) (q: ℕ) (p_prime: prime p) (q_prime: prime q):
```

Comment:
I'm quite shocked this thing is not a fundamental lemma of mathlib or something easily derivable
(1) Are you sure it is not there under an obscure name?
(2) I would suggest to split this into two lemmas corresponding to the two branches of the and
(3) I would suggest to put them into a for_mathlib file for padics

Reply:
(getting the full theorem back is quite easy by and.intro, though the branches have nothing to do together in general)

Reply:
I extracted a lemma and put it in `for_mathlib`.
I still keep this lemma `padic_norm_q` as it handles the interface between `prime` and `nat.prime`.